tensorboard/webapp/webapp_data_source/tb_feature_flag_data_source_test.ts

```diff
@@ -247,5 +247,172 @@ describe('tb_feature_flag_data_source', () => {
+
+        const setItemSpy = spyOn(localStorage, 'setItem').and.stub();
+
+        dataSource.persistFeatureFlags({enabledScalarDataTable: true});
+
+        expect(getPersistentFeatureFlagsSpy).toHaveBeenCalled();
+        expect(setItemSpy).toHaveBeenCalledOnceWith(
+          'tb_feature_flag_storage_key',
+          '{"enabledScalarDataTable":true}'
+        );
+      });
+
+      it('adds new flag when some flags already exist', () => {
+        const getPersistentFeatureFlagsSpy = spyOn(
+          dataSource,
+          'getPersistentFeatureFlags'
+        ).and.callThrough();
+
+        spyOn(localStorage, 'getItem').and.returnValue(
+          '{"enabledScalarDataTable":true}'
+        );
+
+        const setItemSpy = spyOn(localStorage, 'setItem').and.stub();
+
+        dataSource.persistFeatureFlags({enableTimeSeriesPromotion: true});
+
+        expect(getPersistentFeatureFlagsSpy).toHaveBeenCalled();
+        expect(setItemSpy).toHaveBeenCalledOnceWith(
+          'tb_feature_flag_storage_key',
+          '{"enabledScalarDataTable":true,"enableTimeSeriesPromotion":true}'
+        );
+      });
+
+      it('Overrides flag if it is already persisted', () => {
+        const getPersistentFeatureFlagsSpy = spyOn(
+          dataSource,
+          'getPersistentFeatureFlags'
+        ).and.callThrough();
+
+        spyOn(localStorage, 'getItem').and.returnValue(
+          '{"enabledScalarDataTable":true,"enableTimeSeriesPromotion":true}'
+        );
+
+        const setItemSpy = spyOn(localStorage, 'setItem').and.stub();
+
+        dataSource.persistFeatureFlags({enableTimeSeriesPromotion: false});
+
+        expect(getPersistentFeatureFlagsSpy).toHaveBeenCalled();
+        expect(setItemSpy).toHaveBeenCalledOnceWith(
+          'tb_feature_flag_storage_key',
+          '{"enabledScalarDataTable":true,"enableTimeSeriesPromotion":false}'
+        );
+      });
+    });
+
+    describe('getPersistentFeatureFlags', () => {
+      it('returns an empty object when localStorage returns null', () => {
+        const getItemSpy = spyOn(localStorage, 'getItem').and.returnValue(null);
+
+        expect(dataSource.getPersistentFeatureFlags()).toEqual({});
+        expect(getItemSpy).toHaveBeenCalledOnceWith(
+          'tb_feature_flag_storage_key'
+        );
+      });
+
+      it('returns a properly parsed object when getItem gives one', () => {
+        const getItemSpy = spyOn(localStorage, 'getItem').and.returnValue(
+          '{"enabledScalarDataTable":true,"enableTimeSeriesPromotion":false}'
+        );
+
+        expect(dataSource.getPersistentFeatureFlags()).toEqual({
+          enabledScalarDataTable: true,
+          enableTimeSeriesPromotion: false,
+        });
+        expect(getItemSpy).toHaveBeenCalledOnceWith(
+          'tb_feature_flag_storage_key'
+        );
+      });
+    });
+
+    describe('resetPersistentFeatureFlag', () => {
+      it('does nothing there is no flags are persisted', () => {
```

Comment:
nit: 'does nothing there is no' is missing an 'if'.

Reply:
Added.